### ISSUE TITLE
PwaUpdateAvailable button not displayed on page reload

### ISSUE THREAD

**Issue by lpellegr** [thepassle/pwa-helpers#3]
I really like the _PwaUpdateAvailable_ component idea. Unfortunately, the message is displayed only once. Let me explain:

When you deploy a new app version, the browser detects a new service worker and _PwaUpdateAvailable_ reacts by displaying a button to suggest updating the page. At this stage, suppose the user closes the browser tab, the browser, or it's device turns off for whatever reason. Upon page reload, the new service worker remains in _waiting to activate_ state until _skipWaiting_ is called. Unfortunately, the _PwaUpdateAvailable_ component does not redisplay the update button whereas I think it should.

**Comment by thepassle** [comment]
I think this can improved by checking for `ServiceWorkerRegistration.waiting`, ill work on this coming sunday

**Comment by lpellegr** [comment]
If it can help here is an adapted lit-element version I created using `reg.waiting` as suggested:

```javascript
import {LitElement, TemplateResult, html, customElement, CSSResult} from 'lit-element';

import {styles} from './ipregistry-update-available-css';


@customElement('ipregistry-update-available')
export class IpregistryUpdateAvailable extends LitElement {

    private newWorker: ServiceWorker;

    private refreshing: boolean;

    constructor() {
        super();
        this.refreshing = false;
    }

    static get styles(): CSSResult {
        return styles;
    }

    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('click', this.updateApp.bind(this));
        this.setAttribute('hidden', '');

        if ('serviceWorker' in navigator) {
            navigator.serviceWorker.getRegistration().then(reg => {
                reg.addEventListener('updatefound', () => {
                    this.newWorker = reg.installing;
                    this.newWorker.addEventListener('statechange', () => {
                        if (this.newWorker.state === 'installed' && navigator.serviceWorker.controller) {
                            this.dispatchEvent(new CustomEvent('ipregistry-update-available', {detail: true}));
                            this.removeAttribute('hidden');
                        }
                    });
                });

                if (reg.waiting) {
                    this.newWorker = reg.waiting;
                    this.removeAttribute('hidden');
                }
            });

            navigator.serviceWorker.addEventListener('controllerchange', () => {
                if (this.refreshing) {
                    return;
                }

                window.location.reload();
                this.refreshing = true;
            });
        }
    }

    protected render(): TemplateResult | void {
        return html`<div id="container">
                        <p>A new version of the dashboard is available</p>
                        <button @click=${e => this.updateApp}>Update</a>
                    </div>`;
    }

    private updateApp(e) {
        e.preventDefault();

        if (this.newWorker !== undefined) {
            this.sendSkipWaitingMessage();
        }
    }

    private sendSkipWaitingMessage() {
        this.newWorker.postMessage({type: 'SKIP_WAITING'});
    }

}
```

**Comment by thepassle** [comment]
Yeah thats what I was thinking. How is that working for you? Does it show the notification after a refresh?

**Comment by lpellegr** [comment]
Yes, it shows after a refresh.

Le ven. 18 oct. 2019 à 00:08, Pascal Schilp <notifications@github.com> a
écrit :

**Comment by thepassle** [comment]
This is fixed in `0.0.10` by checking for `ServiceWorkerRegistration.waiting`. Thanks for the feedback!
